File: codigos/tps/tp03/TP03Q09_PilhaAlocacaoFlexivelJava/PilhaAlocacaoFlexivel.java
/*
 * Refazer a Questão 6 “Pilha com Alocação Flexível em C".
*/

/*
 * INFO:
 * 
 * - As pilhas são um Tipo Abstrato de Dados (TAD) no qual o primeiro elemento que entra é o último a sair
 *     > First In, Last Out (FILO)
 * 
 * - Tem basicamente os métodos de inserir (empilhar, pop) e remover (desempilhar, push)
 * 
 * - Primeira solução IF e RF
 *     > Por exemplo, inserindo o 1, 3, 5 e 7 e efetuando duas remoções teremos:
 *                     7                               X   X
 *                 5   5                               5   X
 *             3   3   3                               3   3
 *         1   1   1   1                               1   1
 * 
 * - Segunda solução II e RI (inserção e remoção não eficientes)
 *     > Em cada inserção ou remoção, movemos todos os elementos
 *     > Por exemplo, inserindo o 1, 3, 5 e 7 e efetuando duas remoções teremos:
 *                     1                       • Primeira remoção: Retorna o 7 e move todos os demais
 *                 1   3                       • Segunda remoção: Retorna o 5 e move todos os demais
 *             1   3   5                               X   X
 *         1   3   5   7                               1   X
 *                                                     3   1
 *                                                     5   3
 */

/**
 * Pilha com Alocação Flexível em Java
 * @author Bruna Furtado da Fonseca
 * @version 2024-07-16
 */

import java.util.Date;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Scanner;

class Pokemon {
    // Atributos
    private int id;
    private int generation;
    private String name;
    private String description;
    private List<String> types;
    private List<String> abilities;
    private double weight;
    private double height;
    private int captureRate;
    private boolean isLegendary;
    private Date captureDate;

    // Contrutor 1
    public Pokemon() {
        this.id = 0;
        this.generation = 0;
        this.name = null;
        this.description = null;
        this.types = new ArrayList<>();
        this.abilities = new ArrayList<>();
        this.weight = 0;
        this.height = 0;
        this.captureRate = 0;
        this.isLegendary = false;
        this.captureDate = null;
    }

    // Contrutor 2
    public Pokemon(int id, int generation, String name, String description, List<String> types, List<String> abilities,
            double weight, double height, int captureRate, boolean isLegendary, Date captureDate) {
        this.id = id;
        this.generation = generation;
        this.name = name;
        this.description = description;
        this.types = types;
        this.abilities = abilities;
        this.weight = weight;
        this.height = height;
        this.captureRate = captureRate;
        this.isLegendary = isLegendary;
        this.captureDate = captureDate;
    }

    // Métodos get
    public int getId() {
        return this.id;
    }

    public int getGeneration() {
        return this.generation;
    }

    public String getName() {
        return this.name;
    }

    public String getDescription() {
        return this.description;
    }

    public List<String> getTypes() {
        return this.types;
    }

    public List<String> getAbilities() {
        return this.abilities;
    }

    public double getWeight() {
        return this.weight;
    }

    public double getHeight() {
        return this.height;
    }

    public int getCaptureRate() {
        return this.captureRate;
    }

    public boolean getIsLegendary() {
        return this.isLegendary;
    }

    public Date getCaptureDate() {
        return this.captureDate;
    }

    // Métodos set
    public void setId(int id) {
        this.id = id;
    }

    public void setGeneradion(int generation) {
        this.generation = generation;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public void setTypes(List<String> types) {
        this.types = types;
    }

    public void setAbilities(List<String> abilities) {
        this.abilities = abilities;
    }

    public void setWeight(double weight) {
        this.weight = weight;
    }

    public void setHeight(double height) {
        this.height = height;
    }

    public void setCaptureRate(int captureRate) {
        this.captureRate = captureRate;
    }

    public void setIsLegendary(boolean isLegendary) {
        this.isLegendary = isLegendary;
    }

    public void setCaptureDate(Date captureDate) {
        this.captureDate = captureDate;
    }

    /**
     * Lê uma linha do arquivo pokemon.csv e armazena os objetos:
     * id, generation, name, description, types, abilities, weight, height,
     * captureRate,
     * isLegendary e captureDate.
     *
     * @param line String com linha a ser lida
     */
    public void ler(String line) {
        line = line.replaceAll("\"", "");
        String[] lineSplitted = line.split("[\\[\\]]");
        lineSplitted[2] = lineSplitted[2].substring(1);

        // id, generation, name, descripstion
        int comma = lineSplitted[0].indexOf(",");
        for (int i = 0; i < 3; i++)
            comma = lineSplitted[0].indexOf(",", ++comma);

        String subString = lineSplitted[0].substring(0, comma);
        String[] attributes = subString.split(",");

        this.setId(Integer.parseInt(attributes[0]));
        this.setGeneradion(Integer.parseInt(attributes[1]));
        this.setName(attributes[2]);
        this.setDescription(attributes[3]);

        // types
        int start = comma + 1;
        for (int i = 0; i < 2; i++)
            comma = lineSplitted[0].indexOf(",", ++comma);

        subString = lineSplitted[0].substring(start, comma);
        List<String> listAttributes = Arrays.asList(subString.split(","));

        for (int i = 0; i < listAttributes.size(); i++)
            listAttributes.set(i, "'" + listAttributes.get(i) + "'");

        this.setTypes(listAttributes);

        // abilities
        listAttributes = Arrays.asList(lineSplitted[1].split(","));

        for (int i = 0; i < listAttributes.size(); i++)
            listAttributes.set(i, listAttributes.get(i).trim());

        this.setAbilities(listAttributes);

        // weight, height, captureRate, isLegendary, captureDate
        attributes = lineSplitted[2].split(",");

        if (attributes[0] != "") {
            this.setWeight(Double.parseDouble(attributes[0]));
        }
        if (attributes[1] != "") {
            this.setHeight(Double.parseDouble(attributes[1]));
        }
        this.setCaptureRate(Integer.parseInt(attributes[2]));
        this.setIsLegendary(convertBool(attributes[3]));
        this.setCaptureDate(convertDate(attributes[4]));
    }

    /**
     * Converte uma String com um número inteiro 0 ou 1 em tipo boolean
     *
     * @param str String com um número inteiro 0 ou 1
     * @return true / false
     */
    public boolean convertBool(String str) {
        if (str.equals("1"))
            return true;
        else
            return false;
    }

    /**
     * Converte uma String com uma data para o tipo Date
     *
     * @param str String com uma data
     * @return data tipo Date / null
     */
    public Date convertDate(String str) {
        SimpleDateFormat outputFormat = new SimpleDateFormat("dd/MM/yyyy");
        try {
            return outputFormat.parse(str);
        } catch (ParseException e) {
            System.err.println("Erro ao converter a data: " + e.getMessage());
            return null;
        }
    }

    /**
     * Imprime o pokemon selecionado
     */
    public void imprimir() {
        SimpleDateFormat outputFormat = new SimpleDateFormat("dd/MM/yyyy");
        String formattedDate = outputFormat.format(this.getCaptureDate());

        System.out.println(
                "[#" + this.getId() + " -> " + this.getName() + ": " + this.getDescription() + " - " + this.getTypes()
                        + " - " + this.getAbilities() + " - " + this.getWeight() + "kg - " + this.getHeight() + "m - "
                        + this.getCaptureRate()
                        + "% - " + this.getIsLegendary() + " - " + this.getGeneration() + " gen] - " + formattedDate);
    }

    /**
     * Clona o pokemon selecionado
     *
     * @return pokemon clonado
     */
    public Pokemon clone() {
        Pokemon clone = new Pokemon();

        clone.id = this.id;
        clone.generation = this.generation;
        clone.name = this.name;
        clone.description = this.description;
        clone.types = this.types;
        clone.abilities = this.abilities;
        clone.weight = this.weight;
        clone.height = this.height;
        clone.captureRate = this.captureRate;
        clone.isLegendary = this.isLegendary;
        clone.captureDate = this.captureDate;

        return clone;
    }
}

class Celula {
    public Pokemon pokemon; // Elemento inserido na celula.
    public Celula prox; // Aponta a celula prox.

    /**
     * Construtor da classe.
     */
    public Celula() {
        this(null);
    }

    /**
     * Construtor da classe.
     * 
     * @param pokemon pokemon inserido na celula.
     */
    public Celula(Pokemon pokemon) {
        this.pokemon = pokemon;
        this.prox = null;
    }
}

class FlexStack {
    // Atributos
    private Celula top;

    // Construtor 1
    public FlexStack() {
        top = null;
    }

    /**
     * Insere um pokemon na pilha (politica FILO).
     *
     * @param pokemon pokemon a ser inserido
     */
    public void push(Pokemon pokemon) {
        Celula tmp = new Celula(pokemon);
        tmp.prox = top;
        top = tmp;

        tmp = null;
    }

    /**
     * Remove um pokemon da pilha (politica FILO).
     * 
     * @return pokemon removido.
     * @trhows Exception se a sequencia nao conter pokemons.
     */
    Pokemon pop() throws Exception {
        if (top == null)
            throw new Exception("Erro! Não há pokemons para remover.");

        Pokemon removedPokemon = top.pokemon;
        Celula tmp = top;
        top = top.prox;

        tmp.prox = null;
        tmp = null;

        return removedPokemon;
    }

    /**
     * Mostra os pokemon numerados por [index], comecando do topo.
     */
    public void mostrar() {
        int index = 0;
        for (Celula i = top; i != null; i = i.prox) {
            System.out.print("[" + index++ + "] ");
            i.pokemon.imprimir();
        }
    }

    /**
     * Mostra os pokemon numerados por [index], invertendo a impressão. De forma
     * recursiva.
     */
    public void mostraRecursivo() {
        int index = getSizeStack();
        mostraRecursivo(top, index);
    }

    private void mostraRecursivo(Celula i, int index) {
        if (i != null) {
            mostraRecursivo(i.prox, --index);
            System.out.print("[" + index + "] ");
            i.pokemon.imprimir();
        }
    }

    /**
     * Encontra o tamanho da Pilha Flexível.
     */
    public int getSizeStack() {
        int size = 0;
        for (Celula i = top; i != null; i = i.prox, size++)
            ;

        return size;
    }

    /**
     * Soma o peso de todos os pokemon, comecando do topo.
     */
    public double getSumWeight() {
        return getSumWeight(top);
    }

    public double getSumWeight(Celula i) {
        double result = 0;

        if (i != null)
            result += i.pokemon.getWeight() + getSumWeight(i.prox);

        return result;
    }

    /**
     * Soma a altura de todos os pokemon, comecando do topo.
     */
    public double getSumHeight() {
        return getSumHeight(top);
    }

    public double getSumHeight(Celula i) {
        double result = 0;

        if (i != null)
            result += i.pokemon.getHeight() + getSumHeight(i.prox);

        return result;
    }

    /**
     * Encontra o pokemon mais pesado.
     */
    public Pokemon getHeavier() {
        Pokemon heavier = top.pokemon;
        for (Celula i = top.prox; i != null; i = i.prox) {
            if (i.pokemon.getWeight() > heavier.getWeight())
                heavier = i.pokemon;
        }
        return heavier;
    }

    /**
     * Encontra o pokemon mais leve.
     */
    public Pokemon getLighter() {
        Pokemon lighter = top.pokemon;
        for (Celula i = top.prox; i != null; i = i.prox) {
            if (i.pokemon.getWeight() < lighter.getWeight())
                lighter = i.pokemon;
        }
        return lighter;
    }

    /**
     * Encontra o pokemon mais alto.
     */
    public Pokemon getHigher() {
        Pokemon higher = top.pokemon;
        for (Celula i = top.prox; i != null; i = i.prox) {
            if (i.pokemon.getWeight() > higher.getWeight())
                higher = i.pokemon;
        }
        return higher;
    }

    /**
     * Encontra o pokemon mais baixo.
     */
    public Pokemon getLower() {
        Pokemon lower = top.pokemon;
        for (Celula i = top.prox; i != null; i = i.prox) {
            if (i.pokemon.getWeight() < lower.getWeight())
                lower = i.pokemon;
        }
        return lower;
    }
}

public class PilhaAlocacaoFlexivel {
    public static List<Pokemon> pokemonStorage;

    public static Pokemon search(int id) {
        for (Pokemon p : pokemonStorage) {
            if (id == p.getId()) {
                return p;
            }
        }
        return null;
    }

    public static void main(String[] args) throws NumberFormatException, Exception {
        try (BufferedReader br = new BufferedReader(new FileReader("pokemon.csv"))) { // verde.icei.pucminas.br/tmp/pokemon.csv
            String line;
            pokemonStorage = new ArrayList<>();

            br.readLine(); // Lê a primeira linha (cabeçalho)
            while ((line = br.readLine()) != null) {
                Pokemon p = new Pokemon();
                p.ler(line);
                pokemonStorage.add(p);
            }
        } catch (IOException e) {
            System.err.println("Erro ao ler o arquivo: " + e.getMessage());
        }

        Scanner sc = new Scanner(System.in);
        String str;
        FlexStack pokemon = new FlexStack();

        while (!(str = sc.nextLine()).equals("FIM"))
            pokemon.push(search(Integer.parseInt(str)));

        int operations = sc.nextInt(); // Quantidade de registros a serem inseridos/removidos
        sc.nextLine(); // Lê o quebralinha
        int cont = 0;

        while (cont < operations) {
            str = sc.nextLine();

            String[] subString = str.split(" ");

            switch (subString[0]) {
                case "I":
                    pokemon.push(search(Integer.parseInt(subString[1])));
                    break;

                case "R":
                    System.out.println("(R) " + pokemon.pop().getName());
                    break;

                default:
                    break;
            }
            cont++;
        }
        pokemon.mostraRecursivo();
        sc.close();
    }
}

File: codigos/tps/tp03/TP03Q06_PilhaAlocacaoFlexivelC/PilhaAlocacaoFlexivel.c
/*
Pilha com Alocação Flexível em C

author: Bruna Furtado da Fonseca
version: Ubuntu 13.2.0-23ubuntu4
*/

/*
Refazer a Questão 3 “Pilha com Alocação Sequencial".
*/

/*
INFO:

- As pilhas são um Tipo Abstrato de Dados (TAD) no qual o primeiro elemento que entra é o último a sair
    > First In, Last Out (FILO)

- Tem basicamente os métodos de inserir (empilhar, pop) e remover (desempilhar, push)

- Primeira solução IF e RF
    > Por exemplo, inserindo o 1, 3, 5 e 7 e efetuando duas remoções teremos:
                    7                               X   X
                5   5                               5   X
            3   3   3                               3   3
        1   1   1   1                               1   1

- Segunda solução II e RI (inserção e remoção não eficientes)
    > Em cada inserção ou remoção, movemos todos os elementos
    > Por exemplo, inserindo o 1, 3, 5 e 7 e efetuando duas remoções teremos:
                    1                       • Primeira remoção: Retorna o 7 e move todos os demais
                1   3                       • Segunda remoção: Retorna o 5 e move todos os demais
            1   3   5                               X   X
        1   3   5   7                               1   X
                                                    3   1
                                                    5   3
*/

// -----------------------------
// HEADER - Início
// -----------------------------

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Estrutura para armazenar informações do Pokemon
typedef struct Pokemon
{
    int id;
    int generation;
    char *name;
    char *description;
    char **types;
    char **abilities;
    double weight;
    double height;
    int captureRate;
    char *isLegendary;
    char *date;
} Pokemon;

// Estrutura para armazenar um conjunto de Pokémons
typedef struct PokemonStorage
{
    int totalPokemonStorage;
    Pokemon *pokemonStorage;
} PokemonStorage;

// Tipo Celula
typedef struct Celula
{
    Pokemon pokemon;     // Elemento inserido na celula.
    struct Celula *prox; // Aponta a celula prox.
} Celula;

Celula *newCelula(Pokemon pokemon)
{
    Celula *new = (Celula *)malloc(sizeof(Celula));
    new->pokemon = pokemon;
    new->prox = NULL;

    return new;
}

// Definição da estrutura da pilha
typedef struct FlexStack
{
    Celula *top;
} FlexStack;

// Potótipos das funções
void start(PokemonStorage *, FlexStack *);
void ler(PokemonStorage *);
char *str(char *);
Pokemon searchIdStorage(int, PokemonStorage);
void imprimir(Pokemon);
// void clone(int); // NÃO ESTÁ IMPLEMENTADA -----------------------------------------------------

// Potótipos das funções - pilha
void push(FlexStack *, Pokemon);
Pokemon pop(FlexStack *);
void mostrar(FlexStack);
int sizeStack(FlexStack);
void chamaMostrarRecursivo(FlexStack);
void mostrarRecursivo(Celula *, int);
void invertStack(FlexStack *);
void reverseStack(FlexStack *, Pokemon);

// -----------------------------
// HEADER - Fim
// -----------------------------

// FUNÇÃO PRINCIPAL
int main()
{
    PokemonStorage storage;
    FlexStack stack;

    start(&storage, &stack);
    ler(&storage);

    char input[50];

    scanf("%s", input);
    while (strcmp(input, "FIM") != 0)
    {
        push(&stack, searchIdStorage(atoi(input), storage));
        scanf("%s", input);
    }

    int operations; // quantidade de registros a serem inseridos/removidos
    int cont = 0;
    int infos = 0;
    char *pch;
    char subString[2][5];

    scanf("%d", &operations);
    getchar();

    while (cont < operations)
    {
        infos = 0;

        scanf("%[^\n]", input);
        getchar();

        pch = strtok(input, " ");
        while (pch != NULL)
        {
            strcpy(subString[infos], pch); // atribui o tipo tokenizado
            pch = strtok(NULL, " ");
            infos++;
        }

        if (strcmp(subString[0], "I") == 0)
        {
            push(&stack, searchIdStorage(atoi(subString[1]), storage));
        }
        else if (strcmp(subString[0], "R") == 0)
        {
            printf("(R) %s\n", pop(&stack).name);
        }
        cont++;
    }
    chamaMostrarRecursivo(stack);

    // Liberação da memória alocada
    free(storage.pokemonStorage);
    free(stack.top);
}

// Função para inicializar e alocar os atributos
void start(PokemonStorage *storage, FlexStack *stack)
{
    storage->totalPokemonStorage = 0;
    storage->pokemonStorage = (Pokemon *)malloc(1 * sizeof(Pokemon));

    stack->top = NULL;
}

// Função para a leitura do csv
void ler(PokemonStorage *s)
{
    FILE *file = fopen("pokemon.csv", "r");
    if (!file)
    {
        printf("Erro ao abrir o arquivo!\n");
        return;
    }

    char buffer[1000];
    fgets(buffer, sizeof(buffer), file);

    while (fscanf(file, "%d%*c %d%*c", &s->pokemonStorage[s->totalPokemonStorage].id, &s->pokemonStorage[s->totalPokemonStorage].generation) != EOF)
    {                                     // id, generation
        fscanf(file, "%[^,]%*c", buffer); // name
        s->pokemonStorage[s->totalPokemonStorage].name = str(buffer);
        fscanf(file, "%[^,]%*c", buffer); // description
        s->pokemonStorage[s->totalPokemonStorage].description = str(buffer);

        // Types
        int cont = 0;
        char *pch;

        fscanf(file, "%[^\"]%*c", buffer);                                                     // captura os tipos
        s->pokemonStorage[s->totalPokemonStorage].types = (char **)malloc(2 * sizeof(char *)); // aloca espaço inicial para 2 tipos

        pch = strtok(buffer, ","); // Primeira tokenização por vírgula
        while (pch != NULL)
        {
            s->pokemonStorage[s->totalPokemonStorage].types[cont] = str(pch); // atribui o tipo tokenizado
            cont++;
            s->pokemonStorage[s->totalPokemonStorage].types = (char **)realloc(s->pokemonStorage[s->totalPokemonStorage].types, (cont + 1) * sizeof(char *)); // realoca para mais tipos
            pch = strtok(NULL, ",");
        }
        s->pokemonStorage[s->totalPokemonStorage].types[cont] = NULL; // finaliza com NULL

        // Abilities

        fscanf(file, "%*c%*c%[^]]%*c%*c%*c", buffer);                                              // captura as habilidades
        s->pokemonStorage[s->totalPokemonStorage].abilities = (char **)malloc(2 * sizeof(char *)); // aloca espaço inicial para 2 habilidades

        pch = strtok(buffer, "'"); // Primeira tokenização por vírgula

        s->pokemonStorage[s->totalPokemonStorage].abilities[0] = str(pch);
        pch = strtok(NULL, "'");
        pch = strtok(NULL, "'");

        cont = 1;
        while (pch != NULL)
        {
            s->pokemonStorage[s->totalPokemonStorage].abilities[cont] = str(pch);

            cont++;
            s->pokemonStorage[s->totalPokemonStorage].abilities = (char **)realloc(s->pokemonStorage[s->totalPokemonStorage].abilities, (cont + 1) * sizeof(char *)); // realoca para mais habilidades
            pch = strtok(NULL, "'");
            pch = strtok(NULL, "'");
        }
        s->pokemonStorage[s->totalPokemonStorage].abilities[cont] = NULL; // finaliza com NULL

        // weight
        cont = 0;
        int comma = 0;
        while (comma < 1)
        {
            buffer[cont] = fgetc(file);
            if (buffer[cont] == ',')
            {
                comma++;
                s->pokemonStorage[s->totalPokemonStorage].weight = 0.0;
                break;
            }
            else
            {
                cont++;
            }
        }

        if (cont > 1)
        {
            buffer[cont + 1] = '\0';
            s->pokemonStorage[s->totalPokemonStorage].weight = atof(buffer);
        }

        // height
        cont = 0;
        comma = 0;
        while (comma < 1)
        {
            buffer[cont] = fgetc(file);
            if (buffer[cont] == ',')
            {
                comma++;
                s->pokemonStorage[s->totalPokemonStorage].height = 0.0;
                break;
            }
            else
            {
                cont++;
            }
        }

        if (cont > 1)
        {
            buffer[cont + 1] = '\0';
            s->pokemonStorage[s->totalPokemonStorage].height = atof(buffer);
        }

        fscanf(file, "%d%*c", &s->pokemonStorage[s->totalPokemonStorage].captureRate); // captureRate

        fscanf(file, "%[^,]%*c", buffer); // isLegendary
        s->pokemonStorage[s->totalPokemonStorage].isLegendary = strcmp(buffer, "0") == 0 ? str("false") : str("true");

        fscanf(file, "%[^\n]", buffer); // date
        s->pokemonStorage[s->totalPokemonStorage].date = str(buffer);

        s->totalPokemonStorage++;
        s->pokemonStorage = (Pokemon *)realloc(s->pokemonStorage, (s->totalPokemonStorage + 1) * sizeof(Pokemon)); // realoca memória para o próximo Pokémon
    }
    fclose(file);
}

// Função para alocação dinâmica de string
char *str(char *str)
{
    char *info = (char *)malloc((strlen(str) + 1) * sizeof(char));
    strcpy(info, str);
    return info;
}

// Função para pesquisar pokemon por id no armazenamento
Pokemon searchIdStorage(int id, PokemonStorage s)
{
    for (int i = 0; i < s.totalPokemonStorage; ++i)
    {
        if (id == s.pokemonStorage[i].id)
        {
            return s.pokemonStorage[i];
        }
    }
    printf("Pokemon com o id = %d não encontrado no armazenamento\n", id);
    exit(1);
}

// Função para imprimir os pokemons
void imprimir(Pokemon pokemon)
{
    printf("[#%d -> %s: %s - ", pokemon.id, pokemon.name, pokemon.description);

    // Types
    int cont = 0;
    printf("[");
    while (pokemon.types[cont] != NULL)
    {
        if (cont > 0)
            printf(", ");
        printf("'%s'", pokemon.types[cont]);
        cont++;
    }
    printf("] - [");

    // Abilities
    cont = 0;
    while (pokemon.abilities[cont] != NULL)
    {
        if (cont > 0)
            printf(", ");
        printf("'%s'", pokemon.abilities[cont]);
        cont++;
    }
    printf("] - %.1fkg - %.1fm - %d%% - %s - %d gen] - %s\n", pokemon.weight, pokemon.height, pokemon.captureRate, pokemon.isLegendary, pokemon.generation, pokemon.date);
}

// -----------------------------
// PILHA: Início
// -----------------------------

// Função para inserir na pilha
void push(FlexStack *stack, Pokemon pokemon)
{
    Celula *tmp = newCelula(pokemon);
    tmp->prox = stack->top;
    stack->top = tmp;

    tmp = NULL;
}

// Função para remover da pilha
Pokemon pop(FlexStack *stack)
{
    if (stack->top == NULL)
    {
        printf("Erro! Não há pokemons para remover.\n");
        exit(1);
    }

    Pokemon removedPokemon = stack->top->pokemon;
    Celula *tmp = stack->top;
    stack->top = stack->top->prox;

    tmp->prox = NULL;
    free(tmp);

    return removedPokemon;
}

// Função para mostrar os pokemon presentes na pilha
void mostrar(FlexStack stack)
{
    int index = 0;
    for (Celula *i = stack.top; i != NULL; i = i->prox)
    {
        printf("[%d] ", index++);
        imprimir(i->pokemon);
    }
}
// Função para encontrar o tamanho da Pilha Flexível
int sizeStack(FlexStack stack)
{
    Celula *i = stack.top;
    int index;

    for (index = 0; i != NULL; i = i->prox, index++);
    free(i);

    return index;
}

/*
Mostra os pokemon numerados por [index], invertendo a impressão. De forma
recursiva.
*/
void chamaMostrarRecursivo(FlexStack stack)
{
    int index = sizeStack(stack);
    mostrarRecursivo(stack.top, index);
}

void mostrarRecursivo(Celula *i, int index)
{
    if (i != NULL)
    {
        mostrarRecursivo(i->prox, --index);
        printf("[%d] ", index);
        imprimir(i->pokemon);
    }
}

// INVERTER A PILHA

// Função para inserir um elemento (pokemon) no fundo da pilha recursivamente
void reverseStack(FlexStack *stack, Pokemon pokemon)
{
    if (stack->top == NULL)
    {
        push(stack, pokemon);
    }
    else
    {
        Pokemon tmp = pop(stack);
        reverseStack(stack, pokemon);
        push(stack, tmp);
    }
}
/*
Caso Base: Se o topo da pilha (stack->top) for NULL (ou seja, a pilha está vazia), ela apenas empilha (push) o
elemento pokemon.

Caso Recursivo: Caso contrário:
- A função desempilha (pop) o elemento do topo e o armazena na variável tmp.
- Chama reverseStack recursivamente para colocar o pokemon no fundo da pilha.
- Após a recursão, reempilha (push) o elemento tmp que foi desempilhado antes.

Essa lógica permite que a função recorra até o final da pilha, insira o pokemon no fundo e depois remonte a pilha
com os elementos que foram desempilhados.
*/

// Função para inverter a pilha usando a função reverseStack como auxiliar
void invertStack(FlexStack *stack)
{
    if (stack->top != NULL)
    {
        Pokemon tmp = pop(stack);
        invertStack(stack);
        reverseStack(stack, tmp);
    }
}
/*
Caso Base: Se a pilha estiver vazia (stack->top == NULL), a função não faz nada e termina a recursão.

Caso Recursivo:
- A função desempilha (pop) o elemento do topo e o armazena na variável tmp.
- Chama invertStack recursivamente para inverter o restante da pilha.
- Após a recursão, chama reverseStack para inserir o elemento tmp no fundo da pilha invertida.
*/

// -----------------------------
// PILHA - Fim
// -----------------------------